File: guilib-bukkit/src/main/java/net/developertobi/guilib/bukkit/api/pagination/GuiPaginationImpl.java
package net.developertobi.guilib.bukkit.api.pagination;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;
import net.developertobi.guilib.api.gui.GuiController;
import net.developertobi.guilib.api.item.GuiItem;
import net.developertobi.guilib.api.item.GuiPos;
import net.developertobi.guilib.api.pagination.GuiPagination;
import net.developertobi.guilib.bukkit.utils.SoundUtils;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;

public class GuiPaginationImpl implements GuiPagination {

    private final GuiController controller;
    private final List<GuiPos> positions = new ArrayList<>();
    private final Multimap<Integer, GuiItem> items = ArrayListMultimap.create();

    private int currentPageId = 0;
    private int itemsPerPage = 9;

    public GuiPaginationImpl(GuiController controller) {
        this.controller = controller;
    }

    @Override
    public List<GuiPos> getPositions() {
        return positions;
    }

    @Override
    public Multimap<Integer, GuiItem> getItems() {
        return items;
    }

    @Override
    public int getLastPageId() {
        List<Integer> pageIds = new ArrayList<>(items.keySet());
        pageIds.sort(Integer::compareTo);
        return pageIds.isEmpty() ? 0 : pageIds.get(pageIds.size() - 1);
    }

    @Override
    public int getPageAmount() {
        return items.keySet().size();
    }

    @Override
    public int getCurrentPageId() {
        return currentPageId;
    }

    @Override
    public boolean isFirstPage() {
        return currentPageId == 0;
    }

    @Override
    public boolean isLastPage() {
        List<Integer> pageIds = new ArrayList<>(items.keySet());
        pageIds.sort(Integer::compareTo);
        return pageIds.isEmpty() || currentPageId == pageIds.get(pageIds.size() - 1);
    }

    @Override
    public void page(int pageId) {
        List<Integer> pageIds = new ArrayList<>(items.keySet());
        pageIds.sort(Integer::compareTo);
        this.currentPageId = (pageIds.size() <= pageId || pageId < 0) ? 0 : pageId;
        refreshPage();
    }

    @Override
    public void toFirstPage() {
        this.currentPageId = 0;
        refreshPage();
    }

    @Override
    public void toLastPage() {
        if (items.isEmpty()) {
            this.currentPageId = 0;
        } else {
            List<Integer> pageIds = new ArrayList<>(items.keySet());
            pageIds.sort(Integer::compareTo);
            this.currentPageId = pageIds.get(pageIds.size() - 1);
        }
        refreshPage();
    }

    @Override
    public void toNextPage() {
        if (items.isEmpty()) return;
        if (isLastPage()) return;
        this.currentPageId += 1;
        refreshPage();
    }

    @Override
    public void toPreviousPage() {
        if (items.isEmpty()) return;
        if (isFirstPage()) return;
        this.currentPageId -= 1;
        refreshPage();
    }

    @Override
    public void setItemField(int startRow, int startColumn, int endRow, int endColumn) {
        int numRows = controller.getRows();
        int numColumns = controller.getColumns();

        for (int row = startRow; row < Math.min(numRows, endRow + 1); row++) {
            for (int column = startColumn; column < Math.min(numColumns, endColumn + 1); column++) {
                positions.add(GuiPos.of(row, column));
            }
        }
    }

    @Override
    public void distributeItems(List<GuiItem> itemsToDistribute) {
        items.clear();
        int pageIndex = 0;

        for (int i = 0; i < itemsToDistribute.size(); i++) {
            if (i % itemsPerPage == 0) {
                pageIndex = i / itemsPerPage;
            }
            items.put(pageIndex, itemsToDistribute.get(i));
        }

        refreshPage();
    }

    @Override
    public void limitItemsPerPage(int amount) {
        this.itemsPerPage = amount;
    }

    @Override
    public void refreshPage() {
        for (GuiPos currentPosition : positions) {
            if (!controller.isPositionTaken(currentPosition)) continue;
            controller.clearPosition(currentPosition);
        }

        List<GuiItem> itemsForNextPage = new ArrayList<>(items.get(currentPageId));
        int itemIndex = 0;

        for (GuiPos currentPosition : positions) {
            if (itemIndex >= itemsForNextPage.size()) break;
            if (controller.isPositionTaken(currentPosition)) continue;
            controller.setItem(currentPosition, itemsForNextPage.get(itemIndex));
            itemIndex++;
        }

        controller.updateRawInventory();

        if (controller.getProperties().playSoundOnPageSwitch()) {
            Object holder = controller.getRawInventory() != null ? controller.getRawInventory().getHolder() : null;
            if (holder instanceof Player player) {
                SoundUtils.playSwitchPageSound(player);
            }
        }
    }
}

File: guilib-bukkit/bin/main/net/developertobi/inventorylib/bukkit/api/pagination/GuiPaginationImpl.kt
package net.developertobi.inventorylib.bukkit.api.pagination

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.Multimap
import net.developertobi.inventorylib.api.inventory.GuiController
import net.developertobi.inventorylib.api.item.GuiItem
import net.developertobi.inventorylib.api.item.GuiPos
import net.developertobi.inventorylib.api.pagination.GuiPagination
import net.developertobi.inventorylib.bukkit.utils.SoundUtils
import org.bukkit.entity.Player


class GuiPaginationImpl(private val controller: GuiController) : GuiPagination {

    override val positions: MutableList<GuiPos> = mutableListOf()
    override val items: Multimap<Int, GuiItem> = ArrayListMultimap.create()

    private var currentPageId: Int = 0
    private var itemsPerPage: Int = 9

    override fun getLastPageId(): Int {
        val pageIds = items.keySet().stream().toList()
        return pageIds[pageIds.size - 1]
    }

    override fun getPageAmount(): Int {
        return items.keySet().stream().toList().size
    }

    override fun getCurrentPageId(): Int {
        return this.currentPageId
    }

    override fun isFirstPage(): Boolean {
        return this.currentPageId == 0
    }

    override fun isLastPage(): Boolean {
        val pageIds = items.keySet().stream().toList()
        return this.currentPageId == pageIds[pageIds.size - 1]
    }

    override fun page(pageId: Int) {
        val pageIds = items.keySet().stream().toList()
        this.currentPageId = if ((pageIds.size < pageId || pageId < 0)) 0 else pageId
        refreshPage()
    }

    override fun toFirstPage() {
        this.currentPageId = 0
        refreshPage()
    }

    override fun toLastPage() {
        if (items.isEmpty) {
            this.currentPageId = 0
        } else {
            val pageIds = items.keySet().stream().toList()
            this.currentPageId = pageIds[pageIds.size - 1]
        }
        refreshPage()
    }

    override fun toNextPage() {
        if (items.isEmpty) return
        if (isLastPage()) return
        this.currentPageId += 1
        refreshPage()
    }

    override fun toPreviousPage() {
        if (items.isEmpty) return
        if (isFirstPage()) return
        this.currentPageId -= 1
        refreshPage()
    }

    override fun setItemField(startRow: Int, startColumn: Int, endRow: Int, endColumn: Int) {
        val numRows = controller.getRows()
        val numColumns = controller.getColumns()

        for (row in startRow until numRows.coerceAtMost(endRow + 1)) {
            for (column in startColumn until numColumns.coerceAtMost(endColumn + 1)) {
                val position = GuiPos.of(row, column)
                positions.add(position)
            }
        }
    }


    override fun distributeItems(items: List<GuiItem>) {
        this.items.clear()
        var pageIndex = 0

        for (i in items.indices) {
            if (i % itemsPerPage == 0) pageIndex = i / itemsPerPage
            this.items.put(pageIndex, items[i])
        }

        refreshPage()
    }


    override fun limitItemsPerPage(amount: Int) {
        this.itemsPerPage = amount
    }

    override fun refreshPage() {
        for (currentPosition in this.positions) {
            if (!controller.isPositionTaken(currentPosition)) continue
            controller.clearPosition(currentPosition)
        }

        val itemsForNextPage: List<GuiItem> = items[currentPageId].stream().toList()
        var itemIndex = 0

        for (currentPosition in this.positions) {
            if (itemIndex >= itemsForNextPage.size) break
            if (controller.isPositionTaken(currentPosition)) continue
            controller.setItem(currentPosition, itemsForNextPage[itemIndex])
            itemIndex++
        }

        controller.updateRawInventory()

        if (controller.properties.playSoundOnPageSwitch) {
            val holder = controller.rawInventory?.holder ?: return
            SoundUtils.playSwitchPageSound(holder as Player)
        }
    }
}

